core/rpc/client/gateway/jsonrpc/client.go

```diff
@@ -0,0 +1,186 @@
+package jsonrpc
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"sync/atomic"
+
+	rpcClient "github.com/kwilteam/kwil-db/core/rpc/client"
+	"github.com/kwilteam/kwil-db/core/rpc/client/gateway"
+	jsonrpc "github.com/kwilteam/kwil-db/core/rpc/json"
+)
+
+type Client struct {
+	conn     *http.Client
+	endpoint string
+
+	reqID atomic.Int64
+}
+
+// NewClient creates a new gateway json rpc client, target should be the base URL
+// of the gateway server, and should not include "/rpc/v1" as that is appended
+// automatically. If the client does not have a cookie jar, an error is returned.
+func NewClient(target *url.URL, opts ...gateway.ClientOption) (*Client, error) {
+	// This client uses API v1 methods and request/response types.
+	target = target.JoinPath("/rpc/v1")
+
+	c := gateway.DefaultClientOptions()
+	for _, o := range opts {
+		o(c)
+	}
+
+	g := &Client{
+		conn:     c.Client,
+		endpoint: target.String(),
+	}
+
+	// if the caller passed a custom http client without a cookie jar, return an error
+	if g.conn.Jar == nil {
+		return nil, errors.New("gateway http client must have a cookie jar")
+	}
+
+	return g, nil
+}
+
+var _ gateway.Client = (*Client)(nil)
+
+func (g *Client) nextReqID() string {
+	id := g.reqID.Add(-1) // Decrement by 1, so it won't be duplicated with txClient
+	return strconv.FormatInt(id, 10)
+}
+
+// Authn authenticates the client with the gateway.
+// It sets the returned cookie in the client's cookie jar.
+func (g *Client) Authn(ctx context.Context, auth *gateway.AuthnRequest) error {
+	res := &gateway.AuthnResponse{}
+	err := g.call(ctx, gateway.MethodAuthn, auth, res)
+	return err
+}
+
+// GetAuthnParameter returns the auth parameter for the client.
+func (g *Client) GetAuthnParameter(ctx context.Context) (*gateway.AuthnParameterResponse, error) {
+	res := &gateway.AuthnParameterResponse{}
+	err := g.call(ctx, gateway.MethodAuthnParam, &gateway.AuthnParameterRequest{}, res)
+	return res, err
+}
+
+// call sends a JSON-RPC request to the gateway.
+// NOTE: this is basically a copy of core/rpc/client/user/jsonrpc/client.call
+func (cl *Client) call(ctx context.Context, method string, cmd, res any) error {
```

Comment:
In the other client we can export and call this `CallMethod`, so that gateway client can embed it perhaps?   We just want to add the two auth related methods right?

Reply:
yeah I thought about it and totally forgot about it. I think we should do it. I copied bcz that's when we have two different branch , and my branch is based on another branch..

Reply:
Alright I'll make the change